File: day20/day20.py
# !/usr/bin/env python

# Advent of Code 2015
#
# Day 20: Infinite Elves and Infinite Houses
#
# Giacomo Boccardo 2015

import functools
import operator
from itertools import product
from math import sqrt
from timeit import default_timer as timer

from primefac import factorint

input = 36000000
input_div_10 = input / 10
input_div_11 = input / 11
part_2_house_limit = 50


def sqrt_upper_bound(n):
    return set(x for t in ([i, n // i] for i in range(1, int(sqrt(n)) + 1) if n % i == 0) for x in t)


# The factors of odd numbers are always odd, so...
def parity_check(n):
    step = 2 if n % 2 else 1
    return set(x for t in ([i, n // i] for i in range(1, int(sqrt(n)) + 1, step) if n % i == 0) for x in t)


def with_divmod(n):
    result = set()
    for i in range(1, int(sqrt(n)) + 1):
        div, mod = divmod(n, i)
        if mod == 0:
            result |= {i, div}
    return result


def divmod_with_parity_check(n):
    result = set()
    step = 2 if n % 2 else 1
    for i in range(1, int(sqrt(n)) + 1, step):
        div, mod = divmod(n, i)
        if mod == 0:
            result |= {i, div}
    return result


def primefac_library(n):
    primes_and_exps = factorint(n)
    values = [[(f ** e) for e in range(exp + 1)] for (f, exp) in primes_and_exps.iteritems()]
    return set(functools.reduce(operator.mul, v, 1) for v in product(*values))


funcs = [sqrt_upper_bound, parity_check, with_divmod, divmod_with_parity_check, primefac_library]

for f in funcs:
    start = timer()
    house = 1
    part_1 = part_2 = None

    while not (part_1 and part_2):
        house += 1

        divisors = f(house)

        if not part_1 and sum(divisors) >= input_div_10:
            part_1 = house

        if not part_2 and sum(d for d in divisors if house / d <= part_2_house_limit) >= input_div_11:
            part_2 = house

    print("Part 1/2 (using %s): %d/%d (%s)" % (f.__name__, part_1, part_2, timer() - start))
    # 831600/884520

File: day20/day20.swift
import Foundation

// Advent of Code 2015
//
// Day 20: Infinite Elves and Infinite Houses
//
// Giacomo Boccardo 2015

// For the best performances, run from command line using -Ounchecked, i.e.:
//
// $ time swift -Ounchecked day20.swift
// Part 1/2: 831600/884520
//
// real	0m5.025s
// user	0m4.985s
// sys	0m0.035s


func getDivisors(n: Int) -> [Int] {
    var divisors: [Int] = []

    let step = n % 2 == 0 ? 1 : 2

    for i in 1.stride(through: Int(sqrt(Double(n))), by: step) {
        if n % i == 0 {
            divisors.append(i)
            let div = n / i
            if i != div {
                divisors.append(div)
            }
        }
    }

    return divisors
}

func main() {
    let input = 36000000
    let input_div_10 = input / 10
    let input_div_11 = input / 11
    let part_2_house_limit = 50

    var house = 1
    var part_1 = 0
    var part_2 = 0

    while part_1 == 0 || part_2 == 0 {

        let divisors = getDivisors(house)

        if part_1 == 0 && divisors.reduce(0, combine: +) >= input_div_10 {
            part_1 = house
        }

        if part_2 == 0 && divisors.filter({house / $0 <= part_2_house_limit}).reduce(0, combine: +) >= input_div_11 {
            part_2 = house
        }

        house++
    }

    print("Part 1/2: \(part_1)/\(part_2)")

}

main()


